packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/RectangularNode.tsx

```diff
@@ -31,10 +103,109 @@ const labelStyle = (style: Partial<React.CSSProperties>): CSSProperties => {
+
+export const RectangularNode = memo(({ data, isConnectable, id, selected }: NodeProps<RectangularNodeData>) => {
+  const [tools, setTools] = useState<GQLTool[]>([]);
+  const { diagramId, editingContextId } = useContext<DiagramContextValue>(DiagramContext);
+
+  const [getTools, { loading: toolSectionsLoading, data: toolSectionsData, error: toolSectionsError }] = useLazyQuery<
+    GQLGetToolSectionsData,
+    GQLGetToolSectionsVariables
+  >(getToolSectionsQuery);
+
+  useEffect(() => {
+    if (!toolSectionsLoading) {
+      if (toolSectionsData) {
+        const representationDescription: GQLRepresentationDescription =
+          toolSectionsData.viewer.editingContext.representation.description;
+        if (isDiagramDescription(representationDescription)) {
+          const nodeTools = representationDescription.toolSections
+            .flatMap((toolSection) => {
+              return toolSection.tools;
+            })
+            .filter(isSingleClickOnDiagramElementTool);
+          setTools(nodeTools);
+        }
+      }
+    }
+  }, [
+    toolSectionsLoading,
+    toolSectionsData,
+    toolSectionsError,
+    setTools,
+    isSingleClickOnDiagramElementTool,
+    isDiagramDescription,
+  ]);
+
+  useEffect(() => {
+    if (selected) {
+      getTools({
+        variables: {
+          editingContextId: editingContextId,
+          diagramId,
+          diagramElementId: id,
+        },
+      });
+    }
+  }, [selected, editingContextId, diagramId, id, getTools]);
+
+  const [
+    invokeSingleClickOnDiagramElementTool,
+    {
+      loading: invokeSingleClickOnDiagramElementToolLoading,
+      data: invokeSingleClickOnDiagramElementToolData,
+      error: invokeSingleClickOnDiagramElementToolError,
+    },
+  ] = useMutation<GQLInvokeSingleClickOnDiagramElementToolData, GQLInvokeSingleClickOnDiagramElementToolVariables>(
+    invokeSingleClickOnDiagramElementToolMutation
+  );
+
+  useEffect(() => {
+    if (invokeSingleClickOnDiagramElementToolData) {
+      const { invokeSingleClickOnDiagramElementTool } = invokeSingleClickOnDiagramElementToolData;
+      if (invokeSingleClickOnDiagramElementTool.__typename === 'InvokeSingleClickOnDiagramElementToolSuccessPayload') {
+        // TODO: set the selection
+      }
+    }
+  }, [
+    invokeSingleClickOnDiagramElementToolLoading,
+    invokeSingleClickOnDiagramElementToolData,
+    invokeSingleClickOnDiagramElementToolError,
+  ]);
+
+  const invokeTool = (tool: GQLTool) => {
+    if (isSingleClickOnDiagramElementTool(tool)) {
+      const { id: toolId } = tool;
+      const input: GQLInvokeSingleClickOnDiagramElementToolInput = {
+        id: crypto.randomUUID(),
+        editingContextId,
+        representationId: diagramId,
+        diagramElementId: id,
+        toolId,
+        startingPositionX: 0,
+        startingPositionY: 0,
+        selectedObjectId: null,
+      };
+
+      invokeSingleClickOnDiagramElementTool({ variables: { input } });
+    }
+  };
```

Comment:
It should be done in the Palette.
I did try to put that in the Palette.
Unfortunately, the integration of the selection between react flow and the workbench is not done.
Thus, when a tool is applied, the selection is lost.
That makes the Palette close before we receive the payload from the tool mutation making it impossible to update the selection.
So the integration of the selection between react flow and the workbench should be done before.

I will expect the palette not to close when a tool is applied making it possible to set the selection once the tool mutation success payload has been received.

Reply:
Rebased and moved in the palette